alembic/versions/c0810e121564_add_streamflow_order_and_results_tables.py

```diff
@@ -0,0 +1,47 @@
+"""Add streamflow order and results tables
+
+Revision ID: c0810e121564
+Revises: 12f290b63791
+Create Date: 2018-09-05 11:41:58.245456
+
+"""
+from alembic import op
+import sqlalchemy as sa
+
+
+# revision identifiers, used by Alembic.
+revision = 'c0810e121564'
+down_revision = '12f290b63791'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    op.create_table('streamflow_results',
+    sa.Column('streamflow_result_id', sa.Integer(), nullable=False),
+    sa.Column('data_file_id', sa.Integer(), nullable=True),
+    sa.Column('station_id', sa.Integer(), nullable=True),
+    sa.Column('status', sa.Enum('queued', 'processing', 'error', 'cancelled', 'ready', 'removed', name='streamflow_result_statuses'), nullable=False),
+    sa.ForeignKeyConstraint(['data_file_id'], ['data_files.data_file_id'], ),
+    sa.ForeignKeyConstraint(['station_id'], ['stations.station_id'], ),
+    sa.PrimaryKeyConstraint('streamflow_result_id')
+    )
+
+    op.create_table('streamflow_orders',
+    sa.Column('streamflow_order_id', sa.Integer(), nullable=False),
+    sa.Column('hydromodel_output_id', sa.Integer(), nullable=False),
+    sa.Column('streamflow_result_id', sa.Integer(), nullable=False),
+    sa.Column('longitude', sa.Float(), nullable=False),
+    sa.Column('latitude', sa.Float(), nullable=False),
+    sa.Column('notification_method', sa.Enum('none', 'email', name='notification_methods'), nullable=False),
+    sa.Column('notification_address', sa.String(length=255), nullable=True),
+    sa.Column('status', sa.Enum('accepted', 'fulfilled', 'cancelled', 'error', name='streamflow_order_statuses'), nullable=False),
+    sa.ForeignKeyConstraint(['hydromodel_output_id'], ['data_files.data_file_id'], ),
+    sa.ForeignKeyConstraint(['streamflow_result_id'], ['streamflow_results.streamflow_result_id'], ),
+    sa.PrimaryKeyConstraint('streamflow_order_id')
+    )
+
+
+def downgrade():
```

Comment:
This downgrade does not cleanly remove the upgrade, because it does not delete the `streamflow_order_statuses` type. While it's unlikely to be something we need to do, we can't run the upgrade, run the downgrade, and then run the upgrade again without this error message:
```
sqlalchemy.exc.ProgrammingError: (psycopg2.ProgrammingError) type "streamflow_result_statuses" already exists
 [SQL: "CREATE TYPE streamflow_result_statuses AS ENUM ('queued', 'processing', 'error', 'cancelled', 'ready', 'removed')"] (Background on this error at: http://sqlalche.me/e/f405)
```


Is it possible to delete the type?

Reply:
https://www.postgresql.org/docs/9.5/static/sql-droptype.html ?

Reply:
Good catch, @corviday ! 

Hmm. According to SQLAlchemy [docs](https://docs.sqlalchemy.org/en/latest/dialects/postgresql.html#sqlalchemy.dialects.postgresql.ENUM), "The `table.drop()` call will now emit a DROP TYPE for a table-level enumerated type." 

Several things are possible here:

1. You're running the tests with a pre-1.1.0 version of SQLAlchemy.
1. The ENUM isn't table-level. This seems the likeliest to me, but needs some research.
1. The ENUM *is* table-level, but Alembic is doing something that isn't exactly `table.drop()` -- hard to imagine that, but...